ext/zend_test/test.c

```diff
@@ -364,6 +365,67 @@ static ZEND_FUNCTION(zend_test_crash)
 	php_printf("%s", invalid);
 }
 
+zend_mm_heap* zend_test_heap;
+zend_mm_heap* zend_orig_heap;
+volatile uint32_t lineno = 0;
+
+static bool has_opline(zend_execute_data *execute_data)
+{
+	return execute_data
+		&& execute_data->func
+		&& ZEND_USER_CODE(execute_data->func->type)
+		&& execute_data->opline
+	;
+}
+
+#pragma GCC push_options
+#pragma GCC optimize("O0")
+void * __attribute__((optnone)) zend_test_custom_malloc(size_t len)
+{
+	if (has_opline(EG(current_execute_data))) {
+		lineno = EG(current_execute_data)->opline->lineno;
+	}
+	return _zend_mm_alloc(zend_orig_heap, len);
+}
+
+void __attribute__((optnone)) zend_test_custom_free(void *ptr)
+{
+	if (has_opline(EG(current_execute_data))) {
+		lineno = EG(current_execute_data)->opline->lineno;
+	}
+	_zend_mm_free(zend_orig_heap, ptr);
+}
+
+void * __attribute__((optnone)) zend_test_custom_realloc(void * ptr, size_t len)
+{
+	if (has_opline(EG(current_execute_data))) {
+		lineno = EG(current_execute_data)->opline->lineno;
+	}
+	return _zend_mm_realloc(zend_orig_heap, ptr, len);
+}
+#pragma GCC pop_options
+
+static ZEND_FUNCTION(zend_test_observe_opline_in_zendmm)
+{
+	zend_test_heap = malloc(4096);
+	memset(zend_test_heap, 0, 4096);
+	zend_mm_set_custom_handlers(
+		zend_test_heap,
+		zend_test_custom_malloc,
+		zend_test_custom_free,
+		zend_test_custom_realloc
+	);
+	zend_orig_heap = zend_mm_get_heap();
+	zend_mm_set_heap(zend_test_heap);
+}
+
+static ZEND_FUNCTION(zend_test_unobserve_opline_in_zendmm)
+{
+	free(zend_test_heap);
```

Comment:
We have a test that calls every function and checks if ZPP arginfo has problems. So you need `ZEND_PARSE_PARAMETERS_NONE();` to avoid ZPP violation crashes.

Reply:
I switched to an INI setting to en-/disable the opline observing. This also means that we could more easily observe dangling opline pointers in other places without code changes.